apps/homescreen/js/app.js

```diff
@@ -726,6 +757,100 @@
       return this.icons.children.indexOf(child);
     },
 
+    removeSelectedIcon: function() {
+      if (!this.selectedIcon) {
+        return;
+      }
+
+      if (this.selectedIcon.app) {
+        if (!this.selectedIcon.app.removable) {
+          return;
+        }
+        navigator.mozApps.mgmt.uninstall(this.selectedIcon.app);
+      } else if (this.selectedIcon.bookmark) {
+        new MozActivity({
+          name: 'remove-bookmark',
+          data: { type: 'url', url: this.selectedIcon.bookmark.id }
+        });
+      }
+    },
+
+    renameSelectedIcon: function() {
+      if (!this.selectedIcon || !this.selectedIcon.bookmark) {
+        return;
+      }
+
+      new MozActivity({
+        name: 'save-bookmark',
+        data: { type: 'url', url: this.selectedIcon.bookmark.id }
+      });
+    },
+
+    iconIsEditable: function(icon) {
+      return (icon.bookmark || (icon.app && icon.app.removable)) ? true : false;
+    },
+
+    updateSelectedIcon: function(icon) {
+      if (this.selectedIcon === icon) {
+        return;
+      }
+
+      if (this.selectedIcon && (!icon || this.iconIsEditable(icon))) {
+        this.selectedIcon.classList.remove('selected');
+        this.selectedIcon.removeEventListener('touchstart', this);
+        this.selectedIcon = null;
+      }
+
+      var selectedRenameable = false;
+      var selectedRemovable = false;
+
+      if (icon) {
+        selectedRenameable = !!icon.bookmark;
+        selectedRemovable = selectedRenameable || icon.app.removable;
```

Comment:
classList.toggle only works with booleans (it won't remove the class if the evaluation is null or undefined) so probably better to add !!() in here.